src/view/RoomsView.java

```diff
@@ -0,0 +1,129 @@
+
+    JScrollPane scrollPane = new JScrollPane(paneInternals);
+    scrollPane.setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS);
+    scrollPane.setPreferredSize(new Dimension(300, 300));
+
+    rawPane.add(scrollPane);
+
+    messageTextField.addKeyListener(
+        new KeyListener() {
+          @Override
+          public void keyTyped(KeyEvent e) {
+            RoomsState currentState = viewModel.getState();
+            String currentMessage = currentState.getSendMessage();
+            String message = "";
+            if (currentMessage != null) {
+              message = currentMessage;
+            }
+            currentState.setSendMessage(message + e.getKeyChar());
+            viewModel.firePropertyChanged();
+          }
+
+          @Override
+          public void keyPressed(KeyEvent e) {}
+
+          @Override
+          public void keyReleased(KeyEvent e) {}
+        });
+
+    send.addActionListener(
+        evt -> {
+          if (evt.getSource().equals(send)) {
+            RoomsState currentState = viewModel.getState();
+            String message = currentState.getSendMessage();
+            String roomUid = currentState.getRoomUid();
+            if (message != null) {
+              String userUid = currentState.getUserUid();
+              roomsController.sendMessage(roomUid, userUid, message);
+            }
+          }
+        });
+
+    changeRoomButton.addActionListener(
+        evt -> {
+          if (evt.getSource().equals(changeRoomButton)) {
+            RoomsState currentState = viewModel.getState();
+            // This needs to stay like this for now.
+            // Changing rooms will be local to this file
+            // and not involve the controller since we
+            // already have everything here.
+            currentState.setRoomUid("baz");
+            viewModel.firePropertyChanged();
+          }
+        });
+
+    refreshButton.addActionListener(
+        evt -> {
+          if (evt.getSource().equals(refreshButton)) {
+            RoomsState currentState = viewModel.getState();
+            roomsController.loadMessages(currentState.getRoomUid(), currentState.getUserUid());
+          }
+        });
+  }
+
+  @Override
+  public void propertyChange(PropertyChangeEvent evt) {
+    RoomsState state = (RoomsState) evt.getNewValue();
+    String currentRoomUid = state.getRoomUid();
+
+    String roomName = "";
+    for (var room : state.getAvailableRooms()) {
+      if (room.getUid().equals(currentRoomUid)) {
+        roomName = room.getName();
+
+        List<Message> messages = room.getMessages();
+        state.setDisplayMessages(messages);
+
+        paneInternals.removeAll();
+        paneInternals.revalidate();
+        paneInternals.repaint();
+
+        for (var message : messages) {
+          var messageView = new MessageView(message.content);
+          paneInternals.add(messageView);
+        }
+
+        SwingUtilities.invokeLater(
```

Comment:
Why do we use this? (maybe a comment here could be useful)

I'm guessing this is because you managed to solve the weird blank page issue I was running into for a while, which would be very useful! Very strange that you need an InvokeLater though